modules/core/src/main/java/org/apache/ignite/internal/processors/cache/PartitionUpdateCounterErrorWrapper.java

```diff
@@ -0,0 +1,188 @@
+/*
+ * Copyright 2021 GridGain Systems, Inc. and Contributors.
+ *
+ * Licensed under the GridGain Community Edition License (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.gridgain.com/products/software/community-edition/gridgain-community-edition-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.processors.cache;
+
+import org.apache.ignite.IgniteCheckedException;
+import org.apache.ignite.IgniteLogger;
+import org.apache.ignite.internal.util.GridLongList;
+import org.apache.ignite.internal.util.typedef.internal.SB;
+import org.apache.ignite.internal.util.typedef.internal.U;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.Iterator;
+
+/**
+ * Update counter wrapper for error logging.
+ */
+public class PartitionUpdateCounterErrorWrapper implements PartitionUpdateCounter {
+    /** */
+    private final IgniteLogger log;
+
+    /** */
+    private final int partId;
+
+    /** */
+    private final CacheGroupContext grp;
+
+    /** */
+    private final PartitionUpdateCounter delegate;
+
+    /**
+     * @param partId Part id.
+     * @param delegate Delegate.
+     */
+    public PartitionUpdateCounterErrorWrapper(int partId, PartitionUpdateCounter delegate) {
+        this.partId = partId;
+        this.grp = delegate.context();
+        this.log = grp.shared().logger(getClass());
+        this.delegate = delegate;
+    }
+
+    /** {@inheritDoc} */
+    @Override public long reserve(long delta) {
+        String before = toString();
```

Comment:
Why we need before? It seems confusing to prepare string interpretation on every reserve call.